c7n/resources/s3.py

```diff
@@ -2494,3 +2500,119 @@ def process_bucket(self, bucket):
 
         s3.put_bucket_lifecycle_configuration(
             Bucket=bucket['Name'], LifecycleConfiguration={'Rules': config})
+
+
+@actions.register('set-bucket-encryption')
+class SetBucketEncryption(BucketActionBase):
+    """Action enables default encryption on S3 buckets
+
+    `enabled`: boolean Optional: Defaults to True
+    `crypto`: aws:kms | AES256` Optional: Defaults to AES256
+    `key`: arn, alias, or kms id key
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: s3-enable-default-encryption-kms
+                resource: s3
+                actions:
+                  - type: set-bucket-encryption
+                  # enabled: true <------ optional (true by default)
+                    crypto: aws:kms
+                    key: 1234abcd-12ab-34cd-56ef-1234567890ab
+
+              - name: s3-enable-default-encryption-kms-alias
+                resource: s3
+                actions:
+                  - type: set-bucket-encryption
+                  # enabled: true <------ optional (true by default)
+                    crypto: aws:kms
+                    key: alias/some/alias/key
+
+              - name: s3-enable-default-encryption-aes256
+                resource: s3
+                actions:
+                  - type: set-bucket-encryption
+                  # crypto: AES256 <----- optional (AES256 by default)
+                  # enabled: true <------ optional (true by default)
+
+              - name: s3-disable-default-encryption
+                resource: s3
+                actions:
+                  - type: set-bucket-encryption
+                    enabled: false
+    """
+
+    schema = {
+        'type': 'object',
+        'additionalProperties': False,
+        'properties': {
+            'type': {'enum': ['set-bucket-encryption']},
+            'enabled': {'type': 'boolean'},
+            'crypto': {'enum': ['aws:kms', 'AES256']},
+            'key': {'type': 'string'}
+        },
+        'dependencies': {
```

Comment:
hmm hadnt seen this syntax in jsonschema before.

Reply:
schema implementation is based on the `EncryptExtantKeys` action

```
$ cat test.yml
policies:
  - name: s3-encrypt
    resource: s3
    actions:
      - type: set-bucket-encryption
        key: balh
$ custodian validate test.yml                                                                                                         Tue Nov 14 14:41:27 2017
2017-11-14 14:41:34,559: custodian.commands:ERROR Configuration invalid: test.yml
2017-11-14 14:41:34,559: custodian.commands:ERROR u'crypto' is a required property

Failed validating u'required' in schema[6][u'dependencies'][u'key']:
    {u'properties': {u'crypto': {u'pattern': u'aws:kms'}},
     u'required': [u'crypto']}

On instance:
    {'key': 'balh', 'type': 'set-bucket-encryption'}
2017-11-14 14:41:34,559: custodian.commands:ERROR s3-encrypt
```

Reply:
actually i think you have the dependency backwards here your expressing a dependency on the presence of key instead of on the presence of crypto value, ie i could specify crypto: aws:kms without specifying a key here.

https://spacetelescope.github.io/understanding-json-schema/reference/object.html#schema-dependencies

Reply:
Yes, crypto: aws:kms without specifying a key is a valid input for `put_bucket_encryption`, which is kind of weird. However, including a key for crypto: AES256 is not valid.

Reply:
ah ic it uses the default s3 key aws:kms value, sounds good.